### PR TITLE
adds support for including Azure VM metadata

### PR DESCRIPTION
**Pull request by jipperinbham**
### What does this PR do?

 When `collect_azure_tags` is set to true, specific instance metadata will be included as part of the host tags.

### Motivation

Similar functionality already exists for AWS and GCE. My company needs it for Azure as well. 

### Describe your test plan

Used the branch to build a binary and tested on internal infrastructure.

### CODE REVIEW

**Review by sarina-dd — APPROVED** [review]
👍 LGTM, docs

**Comment by albertvaka** [comment]
Hi @jipperinbham! Thanks for implementing this! Note we are currently preparing a release and `master` is frozen, so we can't merge your patch now. Once the release is done, we can get back to this. Thanks again and sorry for the delay this introduces :)

**Comment by jipperinbham** [comment]
Hi @albertvaka, I've now signed the CLA after review from myself and company. Please let me know if there's any changes needed!

File: pkg/util/azure/azure_tags.go [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+	ResourceGroup string `json:"resourceGroupName"`
+}
+
+// GetTags gets the tags from the GCE api
+func GetTags() ([]string, error) {
+	tags := []string{}
+
+	metadataResponse, err := getResponse(metadataURL + "/metadata/instance/compute?api-version=2017-08-01")
+	if err != nil {
+		return tags, fmt.Errorf("unable to query metadata endpoint: %s", err)
```
Comment (hush-hush) [inline pkg/util/azure/azure_tags.go:26]: Could we return `nil` instead of an empty slice for `tags` here ?

File: pkg/util/azure/azure_tags.go [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+	}
+
+	metadata := azureInstanceMetadata{}
+
+	err = json.Unmarshal([]byte(metadataResponse), &metadata)
+	if err != nil {
+		return tags, err
```
Comment (hush-hush) [inline pkg/util/azure/azure_tags.go:33]: Same here.

File: pkg/util/azure/test/azure_metadata.json [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        "name": "exampleosdiskname",
+        "osType": "Linux",
+        "vhd": {
+          "uri": ""
+        },
+        "writeAcceleratorEnabled": "false"
+      }
+  },
+  "subscriptionId": "xxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxx",
+  "tags": "Department:IT;Environment:Test;Role:WebRole",
```
Comment (hush-hush) [inline pkg/util/azure/test/azure_metadata.json:76]: We need to also pull the actual tags set by users.

File: pkg/util/azure/azure_tags.go [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+type azureInstanceMetadata struct {
+	VMID          string `json:"vmId"`
+	Zone          string `json:"zone"`
+	VMSize        string `json:"vmSize"`
+	ResourceGroup string `json:"resourceGroupName"`
+}
+
+// GetTags gets the tags from the GCE api
+func GetTags() ([]string, error) {
```
Comment (hush-hush) [inline pkg/util/azure/azure_tags.go:21]: You will need to check the `cloud_provider_metadata` field like here: https://github.com/DataDog/datadog-agent/blob/master/pkg/util/azure/azure.go#L37

File: releasenotes/notes/collect-azure-metadata-2773087e11627255.yaml [hunk 5]
```diff
@@ -0,0 +1,6 @@
+---
+enhancements:
+  - |
+    Azure VM instance metadata may now be included as host
+    tags by using the ``collect_azure_tags``` configuration
```
Comment (hush-hush) [inline releasenotes/notes/collect-azure-metadata-2773087e11627255.yaml:5]:
```suggestion
    tags by using the `collect_azure_tags` configuration
```

**Review by hush-hush — CHANGES_REQUESTED** [review]
Hi @jipperinbham,

Thanks for this PR. I added a few comments but this looks good overall.

File: pkg/util/azure/azure_tags.go [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+	metadata := azureInstanceMetadata{}
+
+	err = json.Unmarshal([]byte(metadataResponse), &metadata)
+	if err != nil {
+		return tags, err
+	}
+
+	if metadata.VMID != "" {
+		tags = append(tags, fmt.Sprintf("vm-id:%s", metadata.VMID))
+	}
```
Comment (graybjones) [inline pkg/util/azure/azure_tags.go:38]: The crawler does not appear to tag the VM id. It does use it to determine the host name though. The VM id may show up as hostname if no other name is determined.

File: pkg/util/azure/azure_tags.go [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+	if metadata.Zone != "" {
+		tags = append(tags, fmt.Sprintf("zone:%s", metadata.Zone))
+	}
+	if metadata.VMSize != "" {
+		tags = append(tags, fmt.Sprintf("vm-size:%s", metadata.VMSize))
+	}
```
Comment (graybjones) [inline pkg/util/azure/azure_tags.go:44]: The crawler tags vm size as "size"

File: pkg/util/azure/azure_tags.go [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+	if metadata.ResourceGroup != "" {
+		tags = append(tags, fmt.Sprintf("resource-group:%s", metadata.ResourceGroup))
+	}
```
Comment (graybjones) [inline pkg/util/azure/azure_tags.go:47]: resource group is tagged as "resource_group"

File: pkg/util/azure/azure_tags.go [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+	if err != nil {
+		return tags, err
+	}
+
+	if metadata.VMID != "" {
+		tags = append(tags, fmt.Sprintf("vm-id:%s", metadata.VMID))
+	}
+	if metadata.Zone != "" {
+		tags = append(tags, fmt.Sprintf("zone:%s", metadata.Zone))
+	}
```
Comment (graybjones) [inline pkg/util/azure/azure_tags.go:41]: The crawler does not tag zone.

File: pkg/util/azure/azure_tags.go (hunk 6 above)
Comment (jipperinbham) [reply pkg/util/azure/azure_tags.go:38]: @graybjones I'm not familiar with the crawler you're referring to?

File: pkg/util/azure/test/azure_metadata.json (hunk 3 above)
Comment (jipperinbham) [reply pkg/util/azure/test/azure_metadata.json:76]: I wasn't sure whether I should include this initially. Should there be an exclude filter similar to how the [GCE filter](https://github.com/DataDog/datadog-agent/blob/8086c3303dc785d560f1b73e031c4dc3a722ca0a/pkg/util/gce/gce_tags.go#L91-L100) works?

File: pkg/util/azure/azure_tags.go (hunk 6 above)
Comment (graybjones) [reply pkg/util/azure/azure_tags.go:44]: Sorry for the confusion. I will edit my comments for clarity.

File: pkg/util/azure/test/azure_metadata.json (hunk 3 above)
Comment (hush-hush) [reply pkg/util/azure/test/azure_metadata.json:76]: Yes, might be a good idea

**Comment by hush-hush** [comment]
Hi @jipperinbham,

Moving this to 7.23.0. Could you answer @graybjones comments ?

Thanks.

File: pkg/util/azure/azure_tags.go (hunk 7 above)
Comment (jipperinbham) [reply pkg/util/azure/azure_tags.go:50]: Sorry for just not getting back, I've verified we get tags for `resource_group` but we don't seem to be getting one for `size`. Is there a specific setting needed for that to be included in the default host tags?

File: pkg/util/azure/azure_tags.go (hunk 6 above)
Comment (jipperinbham) [reply pkg/util/azure/azure_tags.go:44]: I don't see a way to currently know whether or not the hostname has been set to determine whether or not to include this tag. Any suggestions?

**Pull request closed without merging**